src/heap_utils.cc

```diff
@@ -231,27 +234,220 @@ class BufferOutputStream : public v8::OutputStream {
+    return 65536;  // big chunks == faster
+  }
+
+  void EndOfStream() override {
+    EmitRead(UV_EOF);
+    Cleanup();
+  }
+
+  WriteResult WriteAsciiChunk(char* data, int size) override {
+    int len = size;
+    while (len != 0) {
+      uv_buf_t buf = EmitAlloc(size);
+      ssize_t avail = len;
+      if (static_cast<ssize_t>(buf.len) < avail)
+        avail = buf.len;
+      memcpy(buf.base, data, avail);
+      data += avail;
+      len -= avail;
+      EmitRead(size, buf);
+    }
+    return kContinue;
+  }
+
+  int ReadStart() override {
+    CHECK_NE(snapshot_, nullptr);
+    snapshot_->Serialize(this, HeapSnapshot::kJSON);
+    return 0;
+  }
+
+  int ReadStop() override {
+    return 0;
+  }
+
+  int DoShutdown(ShutdownWrap* req_wrap) override {
+    UNREACHABLE();
+    return 0;
+  }
+
+  int DoWrite(WriteWrap* w,
+              uv_buf_t* bufs,
+              size_t count,
+              uv_stream_t* send_handle) override {
+    UNREACHABLE();
+    return 0;
+  }
+
+  bool IsAlive() override { return snapshot_ != nullptr; }
+  bool IsClosing() override { return snapshot_ == nullptr; }
+  AsyncWrap* GetAsyncWrap() override { return this; }
+
+  void MemoryInfo(MemoryTracker* tracker) const override {
+    if (snapshot_ != nullptr) {
+      tracker->TrackFieldWithSize(
+          "snapshot", sizeof(*snapshot_), "HeapSnapshot");
+    }
+  }
+
+  SET_MEMORY_INFO_NAME(HeapSnapshotStream)
+  SET_SELF_SIZE(HeapSnapshotStream)
+
+ private:
+  void Cleanup() {
+    if (snapshot_ != nullptr) {
+      const_cast<HeapSnapshot*>(snapshot_)->Delete();
+      snapshot_ = nullptr;
+    }
+  }
+
+
+  const HeapSnapshot* snapshot_;
+};
+
+inline void TakeSnapshot(Isolate* isolate, v8::OutputStream* out) {
+  const HeapSnapshot* const snapshot =
+      isolate->GetHeapProfiler()->TakeHeapSnapshot();
+  snapshot->Serialize(out, HeapSnapshot::kJSON);
+  const_cast<HeapSnapshot*>(snapshot)->Delete();
+}
+
+inline bool WriteSnapshot(Isolate* isolate, const char* filename) {
+  FILE* fp = fopen(filename, "w");
+  if (fp == nullptr)
+    return false;
+  FileOutputStream stream(fp);
+  TakeSnapshot(isolate, &stream);
+  fclose(fp);
+  return true;
+}
+
+}  // namespace
+
+void CreateHeapSnapshot(const FunctionCallbackInfo<Value>& args) {
   Isolate* isolate = args.GetIsolate();
-  const HeapSnapshot* snapshot = isolate->GetHeapProfiler()->TakeHeapSnapshot();
   BufferOutputStream out;
-  snapshot->Serialize(&out, HeapSnapshot::kJSON);
-  const_cast<HeapSnapshot*>(snapshot)->Delete();
+  TakeSnapshot(isolate, &out);
   Local<Value> ret;
   if (JSON::Parse(isolate->GetCurrentContext(),
                   out.ToString(isolate)).ToLocal(&ret)) {
     args.GetReturnValue().Set(ret);
   }
 }
 
+void CreateHeapSnapshotStream(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  HandleScope scope(env->isolate());
+  const HeapSnapshot* const snapshot =
+      env->isolate()->GetHeapProfiler()->TakeHeapSnapshot();
+
```

Comment:
```suggestion
CHECK_NOT_NULL(snapshot);
```

Note that it is currently possible when we don't pass ActivityControl, but since we are already being defensive below..